console_win.go

```diff
@@ -496,79 +513,95 @@ func mrec2btns(mbtns, flags uint32) ButtonMask {
 }
 
 func (s *cScreen) getConsoleInput() error {
-	rec := &inputRecord{}
-	var nrec int32
-	rv, _, er := procReadConsoleInput.Call(
-		uintptr(s.in),
-		uintptr(unsafe.Pointer(rec)),
-		uintptr(1),
-		uintptr(unsafe.Pointer(&nrec)))
-	if rv == 0 {
-		return er
-	}
-	if nrec != 1 {
-		return nil
-	}
-	switch rec.typ {
-	case keyEvent:
-		krec := &keyRecord{}
-		krec.isdown = geti32(rec.data[0:])
-		krec.repeat = getu16(rec.data[4:])
-		krec.kcode = getu16(rec.data[6:])
-		krec.scode = getu16(rec.data[8:])
-		krec.ch = getu16(rec.data[10:])
-		krec.mod = getu32(rec.data[12:])
-
-		if krec.isdown == 0 || krec.repeat < 1 {
-			// its a key release event, ignore it
+	waitObjects := []syscall.Handle{s.in, s.cancelflag}
+
+	rv, _, er := procWaitForMultipleObjects.Call(
+		uintptr(len(waitObjects)),
+		uintptr(unsafe.Pointer(&waitObjects[0])),
+		uintptr(0),
+		^uintptr(0))
+	switch rv {
+	case 0: // s.in
```

Comment:
So also  reading the documentation, I don't think the switch statement is entirely correct. Specifically the MS documentation says "If more than one object became signaled during the call, this is the array index of the signaled object with the smallest index value of all the signaled objects."  I read that to mean that its possible for more than one object to be signaled.  So we could have another even more subtle race here.

I'm not sure offhand how we can test whether a specific event is signaled? I think WaitForSingleObject() with a timeout of zero will allow us to query the single event objects state?  Since we are level triggered, this should be safe to use.  (Because you selected manual reset.)